Basic MBP SDF parser

amcastro-tri opened the pull request:
Implements basic MBP parser based off work by @nkoenig in #8441. Geometry parsing will come in a follow up PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8687)
<!-- Reviewable:end -->

amcastro-tri commented:
+@seancurtis-tri for feature review please.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687#-:-LB2J2WC0QjQ1OQRllhM:b-4269ug)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Pass done.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[multibody/benchmarks/acrobot/acrobot.sdf, line 7 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6hK6NC5BQaeV1qy3L:-LB6hK6NC5BQaeV1qy3M:bfi0wn) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/benchmarks/acrobot/acrobot.sdf#L7)):*
> ```
>       <inertial>
>         <mass>1.0</mass>
>         <!-- This inertia is based on a solid cylinder with
> ```

BTW Probably worth putting a comment in here indicating that this sdf file should produce *exactly* the same model as the `MakeAcrobotPlant()` function.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6VVdu4CXJhdiziRaq:-LB6VVdu4CXJhdiziRar:b-sukv49) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L19)):*
> ```objc
> ///
> /// @throws std::runtime_error if the file is not in accordance with the SDF
> /// specification containing a message with a list errors encountered while
> ```

BTW "...a list __*of*__ errors.."

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 21 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6Va38AIFZnHzHZKTZ:-LB6Va38AIFZnHzHZKT_:b-l9lddj) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L21)):*
> ```objc
> /// specification containing a message with a list errors encountered while
> /// parsing the file.
> /// @throws std::logic_error if there is more than one `<model>` element or
> ```

BTW This should probably *also* be a `runtime_error` because you're still accessing a runtime resource from disk.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 15 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6VrjO4_F2ySJGvHlm:-LB6VrjO4_F2ySJGvHln:bv50a9c) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L15)):*
> ```cpp
> 
> using Eigen::Matrix3d;
> using Eigen::Isometry3d;
> ```

FYI alphabetize (swap Isometry and Matrix)

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 29 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6W5cT32DcMZI0B6EO:-LB6W5cT32DcMZI0B6EP:b-bjugst) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L29)):*
> ```cpp
> // Helper function to express an ignition::math::Vector3d instance as
> // a Vector3<double> instance.
> Vector3<double> ToVector3(const ignition::math::Vector3d& vector) {
> ```

BTW You've got a `using Eigen::Vector3d`, but then you use `Vector3<double>` here. Seems odd to use both aliases. Same applies to the `Isometry3<double>` below.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 45 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6WeZ1BPCiGBPExiys:-LB6WeZ1BPCiGBPExiyt:b-qi1txx) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L45)):*
> ```cpp
>     const ignition::math::Inertiald& inertial) {
>   // TODO(amcastro-tri): From MOI() method docs: "Get the moment of inertia
>   // matrix expressed in the base coordinate frame.", What is the "base frame"?
> ```

FYI Seems odd to introduce this TODO in this PR. This isn't a feature, it's about fundamental correctness. So, how can you test this if you don't know what `MOI()`  actually does?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 64 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6X8dLDC66eFrqgLgs:-LB6X8dLDC66eFrqgLgt:b-3t2vf5) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L64)):*
> ```cpp
>   // together with <inertial><pose/></inertial>. It'd seem then the frame B
>   // in X_BI could be another frame. This is currently NOT supported by this
>   // parser.
> ```

BTW It seems that the "NOT supported" functionality, will quietly pass through leading to errors. It would be better to detect and fail fast.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 73 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6XrMa1sMO41hwmO23:-LB6XrMa1sMO41hwmO24:b-3hr39l) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L73)):*
> ```cpp
>   const RotationalInertia<double> I_BBcm_B = I_BBcm_Bi.ReExpress(R_BBi);
> 
>   // Bi's origin is at the COM.
> ```

FYI Is this a documented property of the SDF? If so, I'd indicate that in the comment.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 96 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6Y7UaCgc5H0tP522d:-LB6Y7UaCgc5H0tP522e:b-qir23w) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L96)):*
> ```cpp
>   // Check for any errors.
>   if (!errors.empty()) {
>     std::string error_accumulation("From AddModelFromSdfString():\n");
> ```

BTW This task would be better done with a `std::stringstream`.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 107 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6YTeK5VyxRmHr7fJM:-LB6YTeK5VyxRmHr7fJN:brdekeo) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L107)):*
> ```cpp
> 
>   // Get a pointer to the only model in the file.
>   const sdf::Model* model = root.ModelByIndex(0);
> ```

FYI This line, plus the next, could alternatively be done with:

`const sdf::Model& model = *root.ModelByIndex(0);`

With the added benefit that this version will fail in both release and debug. And, besides, in release mode, the test is skipped and you get the dereferencing error in the for loop. So, it'd all be the same.

I'd apply the same reasoning to all of the pointers you grab in the loop.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 116 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6ZExrB1RiO8bkKF0M:-LB6ZExrB1RiO8bkKF0N:bdg6y99) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L116)):*
> ```cpp
> 
>     // Get the link's inertia relative to the Bcm frame.
>     // Inertial provides a representation for the SpatialInertia M_Bcm_Bi of
> ```

FYI I'd make this `Inertial()` to make it clear you're referring to the method.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 118 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6ZOTZ6cBBUvdtQtAD:-LB6ZOTZ6cBBUvdtQtAE:bdj2xtk) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L118)):*
> ```cpp
>     // Inertial provides a representation for the SpatialInertia M_Bcm_Bi of
>     // body B, about its center of mass Bcm, and expressed in an inertial frame
>     // Bi as defined in <inertial> <pose></pose> </inertial>.
> ```

FYI Given your other comments about the fact that `Bi` may *not* be defined relative `B` but to some other frame, this comment *may* be a lie (and, at least, conflicts with your other comment in `ExtractSpatialInertia()`).

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 137 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6ZpqOAmjBgAFID9rg:-LB6ZpqOAmjBgAFID9rh:b4d66i1) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L137)):*
> ```cpp
>        ++joint_index) {
>     // Get a pointer to the SDF joint, and the joint axis information.
>     const sdf::Joint* joint = model->JointByIndex(joint_index);
> ```

BTW This one doesn't have a `DRAKE_ASSERT`, but I'd still advocate capturing this is a const reference for the reasons previously given.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 152 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6_94rDNQ5OFc0Oqh7:-LB6_94rDNQ5OFc0Oqh8:b-6b9opa) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L152)):*
> ```cpp
>     // And combinations of the above?
>     // There is no way to verify at this level which one is supported or not.
>     // Here we trust that no mather how a user specified the file, joint->Pose()
> ```

FYI s/trust/assume  ("trust" implies that SDF says it's the case, "assume" is the case that SDF didn't say it, but didn't say anything else.)

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 160 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6aCsVAZXG2QxrPacH:-LB6aCsVAZXG2QxrPacI:b67j535) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L160)):*
> ```cpp
>         ToIsometry3(model->LinkByName(joint->ChildLinkName())->Pose());
> 
>     // Compute the location of the child joint in the child link's frame.
> ```

BTW "child joint"?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 166 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6aKTu01UNA53XCtsv:-LB6aKTu01UNA53XCtsw:b-g2kris) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L166)):*
> ```cpp
>     switch (joint->Type()) {
>       case sdf::JointType::REVOLUTE: {
>         // Joint axis, by default in the joint frame J.
> ```

BTW This is a *huge* block for this switch statement. Perhaps it would be better to set the standard now of putting all of this in a function in your anonymous namespace so that future joints follow *that* pattern rather than *this* pattern.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 167 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6b49fAvmfTUaTsULc:-LB6b49fAvmfTUaTsULd:bmlt7yk) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L167)):*
> ```cpp
>       case sdf::JointType::REVOLUTE: {
>         // Joint axis, by default in the joint frame J.
>         // TODO(amcastro-tri): Verify this capability indeed is supported by
> ```

BTW Unclear what "this capability" refers to.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 170 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6bUdmE4J551sz7WeB:-LB6bUdmE4J551sz7WeC:bwf4wyb) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L170)):*
> ```cpp
>         // sdformat.
>         Vector3d axis_J = ToVector3(axis->Xyz());
>         if (axis->UseParentModelFrame()) {
> ```

FYI Do you find "parent model" frame as confusing as I do? A joint is part of a single model. It doesn't make sense to have to add the adjective "parent" to that model? We're not distinguishing it from any other model. Right?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 180 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6bv_xEdamz5f-CmGv:-LB6bv_yEquB4-DWoDRi:b-dccu28) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L180)):*
> ```cpp
>           // TODO(amcastro-tri): Remove these when sdformat guarantees a model
>           // with basic structural checks.
>           if (!model->LinkNameExists(joint->ChildLinkName())) {
> ```

BTW Several things about this test:

1. It seems you've implicitly performed this test by defining `X_MC` above. So, does this test provide any value here?
2. If this validation *needs* to be done, it shouldn't be repeated twice in the "special-case for parent link" code. It should be pulled out and done once -- probably prior to computing `X_MC`.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 185 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6cQrt3LzFg3TeiOs7:-LB6cQrt3LzFg3TeiOs8:b-4mzkfe) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L185)):*
> ```cpp
>           }
> 
>           // If X_PG is left empty, P IS the world frame W, no new frame gets
> ```

BTW s/X_PG/X_PJ

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 187 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6ckZg4x7L3Sta7boM:-LB6ckZg4x7L3Sta7boN:b-88qwdv) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L187)):*
> ```cpp
>           // If X_PG is left empty, P IS the world frame W, no new frame gets
>           // created.
>           optional<Isometry3d> X_PJ;
> ```

BTW This is a subtle point...

You've defined the variable `X_PJ` but this branch is all about the fact that `P == W`. Furthermore, the scope of `X_PJ` is *only* this block. So, this function would be clearer if you basically said, `X_WJ`as the pose of the joint in its parent (world) frame.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 188 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6db2p5wEOasT9dxEM:-LB6db2p5wEOasT9dxEN:b-adv554) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L188)):*
> ```cpp
>           // created.
>           optional<Isometry3d> X_PJ;
>           if (!X_WM.isApprox(Isometry3d::Identity())) X_PJ = X_WM;
> ```

BTW As I see this, if `X_WM` is *not* the identity, then we're going to create a frame with this pose *once* for every pin joint connected to the world? Why not just have the "model" frame?

If the argument is that the odds that there are multiple pin joints in a model connected to the world is vanishingly small, then probably a note to that effect to prevent others from thinking, "hey! Maybe I should optimize this!"

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 195 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6eITg9rXchrRqtos8:-LB6eITg9rXchrRqtos9:b7bnbsq) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L195)):*
> ```cpp
>               plant->GetBodyByName(joint->ChildLinkName()), X_CJ, axis_J);
>         } else {
>           // TODO(amcastro-tri): Remove these when sdformat guarantees a model
> ```

FYI These link-exists functions might well be refactored into a method:

```c++
void ThrowOnMissingLink(const Model& model, const std::string& link_name, const char* link_type) {
  if (!model.LinkNameExists(link_name)) {
    throw std::logic_error("There is no " + link_type + " link named '" + link_name + "' in the model.");
  }
}

ThrowOnMissingLink(joint->ParentLinkName(), "parent");
ThrowOnMissingLink(joint->ChildLinkName(), "child");
```

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 211 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6fC2v5ujf5me6_u89:-LB6fC2v5ujf5me6_u8A:bcegh6q) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L211)):*
> ```cpp
> 
>           // Pose of the frame J in the parent body frame P.
>           const Isometry3d X_PJ = X_MP.inverse() * X_MJ;
> ```

BTW What if `X_PJ` is identity? Why is this not treated the same as connecting to the world frame?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 225 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6fIqT0rWMZue1_FOV:-LB6fIqT0rWMZue1_FOW:babii0) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L225)):*
> ```cpp
>             "Joint type not supported for joint '" + joint->Name() + "'.");
>       }
>     }  // switch
> ```

FYI Is this a thing we do? (Although, by moving the "add joint" into its own function, this block gets small enough that labeling the closing brackets no longer has as much value.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 44 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6fwim1wqt3erJREuB:-LB6fwinCeApQ9uCEOus:btgrlza) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L44)):*
> ```cpp
>     context_ = plant_->CreateDefaultContext();
> 
>     // Create an benchmakr model for verification of the parsed model.
> ```

BTW "a benchmark"

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 92 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6gcgI5EKzJXBZpGdv:-LB6gcgI5EKzJXBZpGdw:b-wfdo7e) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L92)):*
> ```cpp
> TEST_F(AcrobotModelTests, ModelBasics) {
>   // Model Size. Counting the world body, there should be three bodies.
>   EXPECT_EQ(3, plant_->num_bodies());
> ```

BTW Why constants instead of simply referencing the `benchmark_plant_`?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 102 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6giBi8s5NS3Jt0CNl:-LB6giBi8s5NS3Jt0CNm:b-aqz8es) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L102)):*
> ```cpp
>   EXPECT_EQ(plant_->num_multibody_states(), 4);
> 
>   EXPECT_TRUE(plant_->HasJointNamed("ShoulderJoint"));
> ```

BTW Surely this was tested in the `SetUp()` method? (and the next joint).

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 113 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6gvG09uN-Mj-rGzci:-LB6gvG09uN-Mj-rGzcj:b3qkfzl) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L113)):*
> ```cpp
>   // Get joints by name.
>   const Joint<double>& shoulder_joint = plant_->GetJointByName("ShoulderJoint");
>   EXPECT_EQ(shoulder_joint.name(), "ShoulderJoint");
> ```

BTW Can the shoulder joint *possibly* have a name different from the name it was looked up as? That seems to be an invariant of the MBP and testing it here is redundant?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 120 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6h6HX0ktLdaT2nZMS:-LB6h6HX0ktLdaT2nZMT:b-q7al97) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L120)):*
> ```cpp
>   // we just check its index.
>   EXPECT_EQ(shoulder_joint.parent_body().index(), world_index());
>   EXPECT_EQ(shoulder_joint.child_body().name(), "Link1");
> ```

BTW More constants instead of benchmark references?

---

*[tools/workspace/sdformat/package.BUILD.bazel, line 99 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6helm6BD4zoo-ijq1:-LB6helm6BD4zoo-ijq2:bon1vpu) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/tools/workspace/sdformat/package.BUILD.bazel#L99)):*
> ```
>     "include/sdf/Exception.hh",
>     "include/sdf/Filesystem.hh",
>     "include/sdf/Geometry.hh",
> ```

FYI Some of this seems premature for this PR, but probably harmless.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687#-:-LB6hjev0kFfEdUUiIij:b-akjugu)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 45 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6WeZ1BPCiGBPExiys:-LBM6d83EHoi1bmQSPsx:b-1zumyz) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L45)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Seems odd to introduce this TODO in this PR. This isn't a feature, it's about fundamental correctness. So, how can you test this if you don't know what `MOI()`  actually does?
</blockquote></details>

FYI, we are still working on these with @nkoenig. Per latest discussion [here](https://reviewable.io/reviews/robotlocomotion/drake/8441#-L95pd5aA3PVzVDLd07u) it'd seem we are converging towards `R_BBcm_B = Inertial::MOI()` and `R_BBcm_Bi = Inertial::MassMatrix()::MOI()`. However @nkoenig is still working on this and therefore the TODO's as a way to track progress in this regard. For this PR I just went for what I believe is the more stable solution, that is, `R_BBcm_Bi = Inertial::MassMatrix()::MOI()` since it gives you directly (without any transformations under the hood) exactly what was read from the SDF file.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 64 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6X8dLDC66eFrqgLgs:-LBM90m55LT8_-ramyy3:bxpu6zn) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L64)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW It seems that the "NOT supported" functionality, will quietly pass through leading to errors. It would be better to detect and fail fast.
</blockquote></details>

BTW, I misspoke and I clarified in the comment. We do support it as long as sdformat sticks to the contract of always returning X_BI. I know @nkoenig is currently working on this.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 96 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6Y7UaCgc5H0tP522d:-LBMAHfiEEZvIciLoOVc:b1goiko) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L96)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This task would be better done with a `std::stringstream`.
</blockquote></details>

BTW preference or is there a C++ guru reason why that is the case?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 107 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6YTeK5VyxRmHr7fJM:-LBMAot5D5u0fX9Pr3_f:b-v7t31r) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L107)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI This line, plus the next, could alternatively be done with:

`const sdf::Model& model = *root.ModelByIndex(0);`

With the added benefit that this version will fail in both release and debug. And, besides, in release mode, the test is skipped and you get the dereferencing error in the for loop. So, it'd all be the same.

I'd apply the same reasoning to all of the pointers you grab in the loop.
</blockquote></details>

I agree. @nkoenig, probably this API could return references instead of pointers? or is there a reason to keep the pointer semantics? (that is, does the user need to always check for nullptr?)

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 118 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6ZOTZ6cBBUvdtQtAD:-LBMCIWs59XyMGikLuKp:be2kzj6) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L118)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Given your other comments about the fact that `Bi` may *not* be defined relative `B` but to some other frame, this comment *may* be a lie (and, at least, conflicts with your other comment in `ExtractSpatialInertia()`).
</blockquote></details>

I believe I addressed this in your previous comments. Ideally `Inertial` would provide a contract in its API to return these quantities **always** in the same frame, regardless of which of the multiple options to specify frames in an SDF file is used. @nkoenig is working on this and my hope was that the TODO's in this file would help him in this regard.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 152 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6_94rDNQ5OFc0Oqh7:-LBMCv1a-Br0DunuzggV:b-x705wt) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L152)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI s/trust/assume  ("trust" implies that SDF says it's the case, "assume" is the case that SDF didn't say it, but didn't say anything else.)
</blockquote></details>

FYI. Yes, similar issue to the ones you already picked on. SDF offers many ways to specify a frame. @nkoenig is working in this in-memory representation that will satisfy a number of invariants (that is frames in particular) regardless of how exactly things are specified in the SDF file. This TODO is here to document that we rely on this particular invariant.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 166 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6aKTu01UNA53XCtsv:-LBNBTqd4B-VobW54F4M:bmdkg67) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L166)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This is a *huge* block for this switch statement. Perhaps it would be better to set the standard now of putting all of this in a function in your anonymous namespace so that future joints follow *that* pattern rather than *this* pattern.
</blockquote></details>

BTW PTAL.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 170 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6bUdmE4J551sz7WeB:-LBNBnhc-6oi848cPGjQ:b-k3bfl8) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L170)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Do you find "parent model" frame as confusing as I do? A joint is part of a single model. It doesn't make sense to have to add the adjective "parent" to that model? We're not distinguishing it from any other model. Right?
</blockquote></details>

BTW It actually sound ok to me. I do recall though working with Liang that this parent model flag caused some headaches in the past, that's why the TODO to verify it.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 180 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6bv_xEdamz5f-CmGv:-LBNCTsr0wttdAZDjOdJ:b-6a0orm) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L180)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Several things about this test:

1. It seems you've implicitly performed this test by defining `X_MC` above. So, does this test provide any value here?
2. If this validation *needs* to be done, it shouldn't be repeated twice in the "special-case for parent link" code. It should be pulled out and done once -- probably prior to computing `X_MC`.
</blockquote></details>

I think I addressed these with the joint code refactor?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 188 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6db2p5wEOasT9dxEM:-LBND1xg2ZcQn-c_4aj_:b-gmnvwg) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L188)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW As I see this, if `X_WM` is *not* the identity, then we're going to create a frame with this pose *once* for every pin joint connected to the world? Why not just have the "model" frame?

If the argument is that the odds that there are multiple pin joints in a model connected to the world is vanishingly small, then probably a note to that effect to prevent others from thinking, "hey! Maybe I should optimize this!"
</blockquote></details>

I added note in the refactor. Does that help?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 211 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6fC2v5ujf5me6_u89:-LBNDH_FBJ5_3LfEtkR7:bzh9vu5) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L211)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW What if `X_PJ` is identity? Why is this not treated the same as connecting to the world frame?
</blockquote></details>

BTW, good point. The refactor addresses that.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thanks @seancurtis-tri, back to you.

---

Review status: 2 of 6 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687#-:-LBNFbbICCmOUxUtaiuy:bhs38e0)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
+(status: curate commits before merging) 

Next pass done -- one math concern, one strong documentation request, and some typos.

Otherwise: <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> +@sammy-tri for platform review

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[multibody/benchmarks/acrobot/acrobot.sdf, line 4 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBR9JFyE0IGvlyPGbPK:-LBR9JFyE0IGvlyPGbPL:b-furzml) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/benchmarks/acrobot/acrobot.sdf#L4)):*
> ```
>   <model name='acrobot'>
>     <!-- This sdf file produces a model with the default parameters documented
>          in AcrobotParameters in make_acrobot_plant.h. -->
> ```

BTW I feel this language should be stronger. Some kind of warning that it *must* be kept in sync with the default parameters. And a corresponding piece of documentation in `make_acrobot_plant.h` that says if the default parameters change, this file should change as well.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 45 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6WeZ1BPCiGBPExiys:-LBR9vOg4Bm6m8p_2LkQ:b79v8bi) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L45)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

FYI, we are still working on these with @nkoenig. Per latest discussion [here](https://reviewable.io/reviews/robotlocomotion/drake/8441#-L95pd5aA3PVzVDLd07u) it'd seem we are converging towards `R_BBcm_B = Inertial::MOI()` and `R_BBcm_Bi = Inertial::MassMatrix()::MOI()`. However @nkoenig is still working on this and therefore the TODO's as a way to track progress in this regard. For this PR I just went for what I believe is the more stable solution, that is, `R_BBcm_Bi = Inertial::MassMatrix()::MOI()` since it gives you directly (without any transformations under the hood) exactly what was read from the SDF file.
</blockquote></details>

Great rephrasing. I consider those to be much more actionable.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 96 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6Y7UaCgc5H0tP522d:-LBRBxu96StSJ3iZInOh:bv7cgit) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L96)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW preference or is there a C++ guru reason why that is the case?
</blockquote></details>

OK You're building a string. A bunch of `+=` on strings is (possibly) incredibly inefficient. Whereas `stringstream` is designed for exactly that purpose. (There's another more general reason why it exists which is orthogonal to this discussion.)

In a brief chat with @sherm1 this doesn't matter too much. It *may* be the case that this is more expensive, but given that you're doing it right before throwing a program-ending exception, the cost (if it exists) is probably meaningless. And @sherm1 thinks this is prettier.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 152 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6_94rDNQ5OFc0Oqh7:-LBREzDs3QqH44NPrw6v:b-4yxg0j) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L152)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

FYI. Yes, similar issue to the ones you already picked on. SDF offers many ways to specify a frame. @nkoenig is working in this in-memory representation that will satisfy a number of invariants (that is frames in particular) regardless of how exactly things are specified in the SDF file. This TODO is here to document that we rely on this particular invariant.
</blockquote></details>

FYI I would emphasize that "trust" is predicated on something written. Like you did elsewhere, is there a link you could paste here that indicates sdf's published intent?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 75 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRA9PW18gYinS6z0no:-LBRA9PW18gYinS6z0np:b-gnogqg) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L75)):*
> ```cpp
>   // TODO(amcastro-tri): SDF seems to provide <inertial><frame/></inertial>
>   // together with <inertial><pose/></inertial>. It'd seem then the frame B
>   // in X_BI could be another frame. We really on Inertial::Pose() to ALWAYS
> ```

BTW s/really/rely

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 95 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRAztv6DPj0I_u3UWg:-LBRAztv6DPj0I_u3UWh:b-d31jie) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L95)):*
> ```cpp
> }
> 
> // Helper method to retrieve a Body given the name of the linke specification.
> ```

BTW s/linke/link

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 99 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRBKN93anYBoy_VQzd:-LBRBKN93anYBoy_VQze:b-acxmer) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L99)):*
> ```cpp
>     const sdf::Model& model, const std::string& link_name,
>     const MultibodyPlant<double>& plant) {
>   if (link_name.empty() || link_name == "world") {
> ```

BTW This feels like a *very* counter-intuitive overload. Is it really particularly valuable? If so, perhaps documenting this.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 112 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRCkHn9LOxO_1Z9yxp:-LBRCkHn9LOxO_1Z9yxq:b309j13) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L112)):*
BTW "...this helper extracts..." (remove "to")

It might also be better (and consistent with our function documentation as):

> Extracts a Vector3d representation of the joint axis for joints with an axis.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 174 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRHdx49ijV6GgsLHYr:-LBRHdx49ijV6GgsLHYs:b-nv72zx) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L174)):*
> ```cpp
>   optional<Isometry3d> X_PJ;
>   if (parent_body.index() == world_index()) {
>     X_PJ = X_WM;
> ```

BTW I find this confusing.  We're defining `X_PJ`. The test condition suggests that `P = W` and that's fine, but the result is `P = W` *and* `J = M`. However, existence of `X_MJ` suggests that's *not* generally true (otherwise it would simply be the identity).

Generally, I would hope that this special case isn't necessary. That if I were to provide "world" as the link name in the `else` branch, its pose would be the identity and the definition of `X_PJ = X_MP.inverse() * X_MJ;` would be just as correct without special handling.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 48 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRRMu88pCrfAJ_Luf-:-LBRRMu88pCrfAJ_Luf0:b-y47nm8) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L48)):*
> ```cpp
>     context_ = plant_->CreateDefaultContext();
> 
>     // Create an benchmark model for verification of the parsed model.
> ```

BTW s/an/a  (I can't believe I missed that last time)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687#-:-LBR9EZBAWE11vsu-LhC:b8oeyy9)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
First pass complete.  Left a few comments, mostly about documentation an unused includes.

---

Reviewed 2 of 6 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[multibody/multibody_tree/parsing/BUILD.bazel, line 21 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRyjVM0xJJRBUE1QlU:-LBRyjVM0xJJRBUE1QlV:buzc58k) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/BUILD.bazel#L21)):*
> ```Python
>     ],
>     deps = [
>         "//geometry:scene_graph",
> ```

unused?

---

*[multibody/multibody_tree/parsing/BUILD.bazel, line 36 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBS-Z0j69XRz-jzgn5Z:-LBS-Z0j69XRz-jzgn5_:buzc58k) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/BUILD.bazel#L36)):*
> ```Python
>         "//common:find_resource",
>         "//common/test_utilities:eigen_matrix_compare",
>         "//common/test_utilities:expect_throws_message",
> ```

unused?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 3 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRx5QG7stGf7f7s408:-LBRx5QG7stGf7f7s409:b-d4n1o3) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L3)):*
> ```objc
> #pragma once
> 
> #include <memory>
> ```

unused in the header?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 6 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRx6kY5H6v2bdLC7xV:-LBRx6kY5H6v2bdLC7xW:buzc58k) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L6)):*
> ```objc
> #include <string>
> 
> #include "drake/geometry/scene_graph.h"
> ```

unused?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 157 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRyG8S0nT7P7Au0r1_:-LBRyG8S0nT7P7Au0r1a:b-tscwd8) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L157)):*
> ```cpp
>   //  - <joint> <pose> </pose></joint>.
>   //  - <joint> <pose> <frame/> </pose></joint>.
>   //  - <joint> <frame><pose> <frame/> </pose></frame> </pose></joint>.
> ```

BTW this seems like one closing pose tag too many?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 9 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRywxO9CxpgcN1nZep:-LBRywxO9CxpgcN1nZeq:buzc58k) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L9)):*
> ```cpp
> #include "drake/common/find_resource.h"
> #include "drake/common/test_utilities/eigen_matrix_compare.h"
> #include "drake/geometry/scene_graph.h"
> ```

unused?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 19 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRyz8pE985TbdgRC9G:-LBRyz8pE985TbdgRC9H:buzc58k) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L19)):*
> ```cpp
> 
> using Eigen::Vector3d;
> using geometry::SceneGraph;
> ```

unused?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 24 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRzWOj4jY4bk8PpSYx:-LBRzWOj4jY4bk8PpSYy:buzc58k) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L24)):*
> ```cpp
> using multibody::parsing::AddModelFromSdfFile;
> using systems::Context;
> using systems::LeafSystem;
> ```

unused?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 98 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRzsodCEZeT6oJqNl4:-LBRzsodCEZeT6oJqNl5:bdg6w89) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L98)):*
> ```cpp
>   EXPECT_EQ(benchmark_plant_->num_bodies(), plant_->num_bodies());
>   EXPECT_EQ(benchmark_plant_->num_joints(), plant_->num_joints());
>   // Even though our benchmark model does, the parsed model has not actuators.
> ```

typo "has no actuators"?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 120 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBS--Z66F7CBnwyJZ4_:-LBS--Z66F7CBnwyJZ4a:beqswck) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L120)):*
> ```cpp
>   EXPECT_EQ(elbow_joint.name(), "ElbowJoint");
> 
>   // Drake uses a different convention for naming the world body and therefore
> ```

We...  kinda are Drake too?  I'm a little unclear on why there are different conventions for naming the world body which are both considered valid instances of MultibodyPlant and I'm not sure what this comment is telling me about thast.

---

*[tools/workspace/sdformat/package.BUILD.bazel, line 156 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRzCkQByxjR3_aA51d:-LBRzCkS2MooHZxsSzA1:bjg8qgz) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/tools/workspace/sdformat/package.BUILD.bazel#L156)):*
> ```
>         "src/Filesystem.cc",
>         "src/Geometry.cc",
>         "src/Joint.cc",
> ```

BTW inconsistent ordering of Joint vs JointAxis here vs. headers

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687#-:-LBS0R1v2aHzdVmKb1oq:b-r0gg0g)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
-(status: skip ci) 

---

Review status: all files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687#-:-LBS0h-7Ad9Meih60JWv:b-kdkf7e)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Another mindless test +(status: skip ci) 

---

Review status: all files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687#-:-LBS0khOB6FJyE-zAtHD:bfa4lzi)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[multibody/multibody_tree/parsing/BUILD.bazel, line 21 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRyjVM0xJJRBUE1QlU:-LBSOw4P8BsUHTZ0CmDm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/BUILD.bazel#L21)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/BUILD.bazel, line 36 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBS-Z0j69XRz-jzgn5Z:-LBSP4Eh6Ehdf0FznsF4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/BUILD.bazel#L36)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 3 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRx5QG7stGf7f7s408:-LBSPLPBBPrtN7f03XSA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L3)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

unused in the header?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 6 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRx6kY5H6v2bdLC7xV:-LBSPSf26ewrOGmsZKsU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L6)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 96 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6Y7UaCgc5H0tP522d:-LBSQPAS5pJCw1o_s1L1:byaoa9k) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L96)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

OK You're building a string. A bunch of `+=` on strings is (possibly) incredibly inefficient. Whereas `stringstream` is designed for exactly that purpose. (There's another more general reason why it exists which is orthogonal to this discussion.)

In a brief chat with @sherm1 this doesn't matter too much. It *may* be the case that this is more expensive, but given that you're doing it right before throwing a program-ending exception, the cost (if it exists) is probably meaningless. And @sherm1 thinks this is prettier.
</blockquote></details>

BTW pretty is good!

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 152 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LB6_94rDNQ5OFc0Oqh7:-LBSR3Dd3pfUp9ufx8x1:b294kxu) ([raw file](https://github.com/robotlocomotion/drake/blob/e7882e4af40c494835ab212eb2685c36e9564886/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L152)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I would emphasize that "trust" is predicated on something written. Like you did elsewhere, is there a link you could paste here that indicates sdf's published intent?
</blockquote></details>

I wish I could find that link, but nope. It's all based on previous experience. I believe @nkoenig was actually working on a frame abstraction so that you would be able to query the pose of any frame in any other frame. However that work is not done yet and we can only rely on SDF keeping this invariant (which we know from experience). Maybe @nkoenig can clarify?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 157 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRyG8S0nT7P7Au0r1_:-LBSQoLw4_HI4cUUly72:bls7fae) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L157)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW this seems like one closing pose tag too many?
</blockquote></details>

wow! you are a living parser! how did you even spot that?! Thanks

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 9 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRywxO9CxpgcN1nZep:-LBSSd7J4y1gyU8qoxdG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L9)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 19 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRyz8pE985TbdgRC9G:-LBSSfD65QQ1aZ9Bowd8:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L19)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 24 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRzWOj4jY4bk8PpSYx:-LBSShGk2a8YYXANpvvZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L24)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 48 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRRMu88pCrfAJ_Luf-:-LBSSqJR6Qpta_8VxsKN:b-5gt69z) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L48)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW s/an/a  (I can't believe I missed that last time)
</blockquote></details>

you just redeemed yourself then. 

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 98 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRzsodCEZeT6oJqNl4:-LBSSyM0AooKGIWmFD8X:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L98)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

typo "has no actuators"?
</blockquote></details>

Done.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 120 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBS--Z66F7CBnwyJZ4_:-LBSTCvy4B0UHC5_Lt2U:bvlsf3m) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L120)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

We...  kinda are Drake too?  I'm a little unclear on why there are different conventions for naming the world body which are both considered valid instances of MultibodyPlant and I'm not sure what this comment is telling me about thast.
</blockquote></details>

Never mind. I guess that leaked from a previous iteration and didn't make it through. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc, line 174 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRHdx49ijV6GgsLHYr:-LBSnOPFBxy40GR0BzkQ:b-40bbk8) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.cc#L174)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I find this confusing.  We're defining `X_PJ`. The test condition suggests that `P = W` and that's fine, but the result is `P = W` *and* `J = M`. However, existence of `X_MJ` suggests that's *not* generally true (otherwise it would simply be the identity).

Generally, I would hope that this special case isn't necessary. That if I were to provide "world" as the link name in the `else` branch, its pose would be the identity and the definition of `X_PJ = X_MP.inverse() * X_MJ;` would be just as correct without special handling.
</blockquote></details>

I believe there's no need to assume `J = M` even in this case. The parent body can be the world but the joint frame could still have a "fixed" offset in that world.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 24 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8687#-LBRzWOj4jY4bk8PpSYx:-LBSnJ-qBsppvzYxzcv3:bfn1624) ([raw file](https://github.com/robotlocomotion/drake/blob/7be57df39330d0f4feac19837c467675215ff179/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L24)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

the using for `LeafSystem` still seems to be present?
</blockquote></details>

sorry, I removed the header and forgot the `using`. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8687)*
<!-- Sent from Reviewable.io -->
